org.openscenegraph.android/jni/JNIosgUtil.cpp

```diff
@@ -128,4 +180,284 @@ JNIEXPORT jlong JNICALL Java_org_openscenegraph_osg_util_GeometryUtils_nativeCre
+	if(_geode==NULL)
+	{
+		LOGE("Invalid Geode.");
+		return 0;
+	}
+	if(_geode->getNumDrawables()<1)
+	{
+		LOGE("Geode has no drawables");
+		return 0;
+	}
+	osg::Geometry* _geometry = reinterpret_cast<osg::Geometry*>(_geode->getDrawable(_geode->getNumDrawables()-1));
+	if(_geometry==NULL)
+	{
+		LOGE("Invalid Geometry.");
+		return 0;
+	}
+
+	int colored_point = 0;
+	bool use_texcoord = false, new_color_array = false;
+
+	osg::Vec3Array* varray = (osg::Vec3Array*)(_geometry->getVertexArray());
+	if(varray==NULL)
+	{
+		LOGE("Invalid VertexArray");
+		return 0;
+	}
+
+	osg::Vec4Array* carray = (osg::Vec4Array*)(_geometry->getColorArray());
+	if(carray==NULL)
+	{
+		carray = new osg::Vec4Array();
+		carray->ref();
+		new_color_array = true;
+	}
+	if(carray->getNumElements() < varray->getNumElements())
+	{
+		carray->resizeArray(varray->getNumElements());
+	}
+	osg::Vec2Array* tarray = (osg::Vec2Array*)(_geometry->getTexCoordArray(0));
+	if(tarray!=NULL)
+	{
+		use_texcoord = true;
+		if(tarray->getNumElements() < varray->getNumElements())
+		{
+			tarray->resizeArray(varray->getNumElements());
+		}
+	}
+
+	osg::Image* img = osgDB::readImageFile(jstring2string(env,img_filename));
+	if(img==NULL)
+	{
+		LOGE("Invalid Image");
+		return 0;
+	}
+
+	RefVec3* Cg = reinterpret_cast<RefVec3 *>(Cg_ptr);
+	RefVec3* R = reinterpret_cast<RefVec3 *>(R_ptr);
+	osg::RefMatrixf* trmat = reinterpret_cast<osg::RefMatrixf* >(trmat_ptr);
+	osg::Matrixf rotmat;
+	rotmat.set((*trmat)(0,0), (*trmat)(0,1), (*trmat)(0,2), 0, (*trmat)(1,0), (*trmat)(1,1), (*trmat)(1,2), 0, (*trmat)(2,0), (*trmat)(2,1), (*trmat)(2,2), 0, 0, 0, 0, 1);
+
+	//std::stringstream sstrm;
+	//sstrm << "Mat: " << rotmat(0,0) << ", " << rotmat(0,1) << ", " << rotmat(0,2) << std::endl;
+	//sstrm << rotmat(1,0) << ", " << rotmat(1,1) << ", " << rotmat(1,2) << std::endl;
+	//sstrm << rotmat(2,0) << ", " << rotmat(2,1) << ", " << rotmat(2,2) << std::endl;
+	//sstrm << "Cg: " << Cg->x() << ", " << Cg->y() << ", " << Cg->z() << std::endl;
+	//__android_log_write(ANDROID_LOG_INFO,LOG_TAG,sstrm.str().c_str());
+
+	osg::Vec3 M, D;
+	for(unsigned int i = 0; i < varray->getNumElements(); i++)
+	{
+		M = varray->at(i);
+		D = rotmat * (M-(*Cg));
+		if(D.z()>0)
+		{
+			float u = ((R->z()*(D.x()/D.z()))+(R->x()/2.0f))/R->x();
+			float v = ((R->z()*(D.y()/D.z()))+(R->y()/2.0f))/R->y();
+			if( (u>0) && (u<1) && (v>0) && (v<1) )
+			{
+				osg::Vec2 tex_coord(u,v);
+				if(use_texcoord)
+				{
+					tarray->at((unsigned int)i) = tex_coord;
+				}
+				osg::Vec4 _color = Interpolate(tex_coord, img);
+				carray->at((unsigned int)i) = _color;
+				colored_point++;
+			}
+		}
+	}
+
+	if(new_color_array)
+	{
+		_geometry->setColorArray(carray);
+		_geometry->setColorBinding(osg::Geometry::BIND_PER_VERTEX);
+	}
+	if(use_texcoord)
+	{
+		_geometry->setTexCoordArray(0,tarray);
+	}
+
+	osgDB::writeNodeFile(*_node, jstring2string(env,out_filepath));
+	return reinterpret_cast<jint>(colored_point);
+}
```

Comment:
The functions colour(s) a point/set of points using a given image texture that is connected with a camera pose facing the geometry (in world coordinate system). Thus, the texture coordinate parameterization established in the mid is just valid for a given camera position (which is expressed as a transformation matrix of the camera) the function determines the local image coordinate for the/each point and, if within the camera frame, colours the point accordingly.
To the parameters: trmat is the camera matrix, R is the intrinsic camera matrix (holding the frame extend in metres and the focal length), and Cg is the camera center vector.